sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala

```diff
@@ -65,6 +65,6 @@ object Canonicalize {
       val newChildren = orderCommutative(l, { case Least(children) => children })
       Least(newChildren)
 
-    case _ => e.withNewChildren(e.children.map(reorderCommutativeOperators))
+    case _ => e
```

Comment:
I have some doubts regarding this change. 

What about a tree where a commutative child has a non-commutative parent which is not listed above for example `abs('a + 'b)`? As with this change the recursive call stops at `abs`...

can you extend the test cases which such an example?

Reply:
That should not be an issue. The children traversal is not needed, because the precanonicalize on children would ensure that all the elements of subtree have the precanonicalize invoked, which would mean that any commutative expression's precanonicalize is also invoked, and for commutative expressions, the precanonicalize & canonicalize are identical & involves reordering.

I will add tests for that too,.

Also I think now there is no need for precanonicalize variable at all.  whatever code is in precanonicalize can directly be  assigned to canonicalize variable.
The main issue was  tree traversal at each level ( & mostly redundant) caused by reorderCall.  that is taken care of by not dwelving into children and reorder is being called only for commutative expressions.
I will also measure the perf impact. that should not change.
That will simplify the code too.

Reply:
@attilapiros  btw just realized that the case you mentioned as such is what is being tested in the bug tests as GreaterThan , EqualTo etc are non commutative expressions.. 
But I will use abs & also have a deeply nested expression tree containing a mix of commutative & non commutative expressions.